### PR TITLE
feat: add a new package: rn-changelog-generator

### PR DESCRIPTION
**Pull request by kelset**
### Description

This is a new package being added to the rnx-kit suite; it's meant to pack just one functionality exposed as a CLI, to generate the changelog for react-native core. It used to live here https://github.com/react-native-community/releases/blob/master/scripts/changelog-generator.ts but since we are deprecating that repo, we wanted a place to put it while we do some more substantial rework for how the changelog is implemented for RN.

This PRs aims to just get the ball rolling, creating an npm package that can be used via `npx` so that @lunaleaps can start using it for her CI works on RN. The way the command is used is as such:

```
yarn generate --base v0.65.0 --compare v0.66.0 --repo ../../../react-native --changelog ../../../react-native/CHANGELOG.md --token [GH_TOKEN] > NEW_CHANGES.md
```

*(yes, it has great margins of improvements)*

We are putting it here to leverage the existing great infra & CI present in rnx-kit, since it's Typescript based and RN's core is Flow-based.

### Test plan

We are not able to run the existing test suite on CI because it requires the presence of the react-native repo locally; because of that, I've renamed the `test` script to `test:local`. It needs to be ran like so:

```sh
RN_REPO="../../../react-native" yarn test:local
```

### CODE REVIEW

File: packages/rn-changelog-generator/package.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    "url": "https://github.com/microsoft/rnx-kit",
+    "directory": "packages/rn-changelog-generator"
+  },
+  "engines": {
+    "node": ">= 14"
+  },
+  "scripts": {
+    "build": "rnx-kit-scripts build",
+    "format": "rnx-kit-scripts format",
+    "generate": "env TS_NODE_TRANSPILE_ONLY=true node $_NODE_OPTS -r ts-node/register ./lib/changelog-generator.js",
```
Comment (tido64) [inline packages/rn-changelog-generator/package.json:27]: What are we using ts-node for? It looks like we consume a plain .js file?

File: packages/rn-changelog-generator/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+<!-- We recommend that the first beachball change report for a new package is of type "none" -->
+
```
Comment (tido64) [inline packages/rn-changelog-generator/README.md:2]: Remove?
```suggestion
```

File: packages/rn-changelog-generator/package.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  },
+  "devDependencies": {
+    "@octokit/rest": "^18.5.3",
+    "@rnx-kit/scripts": "*",
+    "@types/fast-levenshtein": "0.0.2",
+    "chalk": "^4.1.0",
+    "deepmerge": "^4.2.2",
+    "p-limit": "^3.1.0",
+    "fast-levenshtein": "^3.0.0",
+    "simple-git": "^2.39.0",
```
Comment (tido64) [inline packages/rn-changelog-generator/package.json:38]: Where are we using simple-git?

Comment (kelset) [reply packages/rn-changelog-generator/package.json:38]: good catch! It seems it's not used anywhere

Comment (kelset) [reply packages/rn-changelog-generator/package.json:38]: it was present in the other script (which doesn't need to be ported: https://github.com/react-native-community/releases/blob/master/scripts/cherry-pick.js)

**Review by tido64 — APPROVED** [review]

File: packages/rn-changelog-generator/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# React Native core changelog generator
+
+[![Build](https://github.com/microsoft/rnx-kit/actions/workflows/build.yml/badge.svg)](https://github.com/microsoft/rnx-kit/actions/workflows/build.yml)
+[![npm version](https://img.shields.io/npm/v/@rnx-kit/rn-changelog-generator)](https://www.npmjs.com/package/@rnx-kit/rn-changelog-generator)
+
+This script used to live in the
+[RNC/releases repo](https://github.com/react-native-community/releases/blob/master/scripts/changelog-generator.ts)
+but since we are deprecating it, we wanted to keep the script alive somewhere
+where a good infra was present, so here we are!
```
Comment (afoxman) [inline packages/rn-changelog-generator/README.md:9]:
```suggestion
This script used to live in
[react-native-community/releases](https://github.com/react-native-community/releases/blob/master/scripts/changelog-generator.ts), and has moved here as part of deprecating that repo.
```

File: packages/rn-changelog-generator/README.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+Then create a GitHub ‘personal access token’ that can be used by the script to
+fetch commit metadata from the GitHub API.
+
+1. Visit the [token settings page](https://github.com/settings/tokens).
+1. Generate a new token and **only** give it the `public_repo` scope.
+1. Store this token somewhere secure for future usage.
+
+## Usage
+
+Simply run:
```
Comment (afoxman) [inline packages/rn-changelog-generator/README.md:74]:
```suggestion
Generate a changelog for `react-native` commits between versions 0.65.0 and 0.66.0:
```

File: packages/rn-changelog-generator/README.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+```sh
+npx rn-changelog-generator --base v0.65.0 --compare v0.66.0 --repo ../../../react-native --changelog ../../../react-native/CHANGELOG.md --token [GH_TOKEN] > NEW_CHANGES.md
+```
+
+As explained above, you will need to have a local clone of the react-native core
+repo to refer in the command to make it work (along with the GH token).
```
Comment (afoxman) [inline packages/rn-changelog-generator/README.md:81]:
```suggestion
As explained above, you will need to have a local clone of `react-native`, which is referenced by the `--repo` parameter. You'll also need to provide a GitHub personal access token for the `--token` parameter.
```

File: packages/rn-changelog-generator/package.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "@rnx-kit/rn-changelog-generator",
+  "version": "0.0.1",
+  "description": "New package called rn-changelog-generator",
```
Comment (afoxman) [inline packages/rn-changelog-generator/package.json:4]:
```suggestion
  "description": "Generate a markdown-formatted list of changes between two version of react-native",
```

File: packages/rn-changelog-generator/src/changelog-generator.ts [hunk 8]
```diff
@@ -0,0 +1,838 @@
+#!/usr/bin/env node
```
Comment (afoxman) [inline packages/rn-changelog-generator/src/changelog-generator.ts:1]: I'm skimming this file, holding off on most comments for now, as I know this is just to get the existing code secured in our repo and establish a baseline for it. Down the road, I'd like to see the code split up into logical units, each in their own file. And the deeply-nested call chaining is hard to follow - maybe separate those out into separate functions.

File: packages/rn-changelog-generator/test/changelog-generator.test.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+  getChangeMessage,
+  getChangelogDesc,
+  getOffsetBaseCommit,
+  getOriginalCommit,
+  getFirstCommitAfterForkingFromMain,
+  Changes,
+  PlatformChanges,
+} from "../src/changelog-generator";
+
+if (!process.env.RN_REPO) {
```
Comment (afoxman) [inline packages/rn-changelog-generator/test/changelog-generator.test.ts:25]: When you work on refactoring things, I'd like to see most tests move from "local" (manual) to automated. I'm thinking of things like validating individual functions that do work but don't call into git. This may require some refactoring to separate the code a bit more.

You'll still want an end-to-end integration test suite like you have here. But the more you can move into automated testing, the higher the quality will remain.

**Review by afoxman — COMMENTED** [review]
Most of my comments were around wording (nits) and future work I'd like to see once you get this checked in to establish as baseline for the script.

File: packages/rn-changelog-generator/src/changelog-generator.ts (hunk 8 above)
Comment (kelset) [reply packages/rn-changelog-generator/src/changelog-generator.ts:1]: Yeah I agree - I also want to refactor it.

File: packages/rn-changelog-generator/README.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  For this reason, the script will parse the ‘referential revision’–which is
+  added to each commit message by Facebook’s infrastructure–and use it to
+  resolve to the original commit in the `main` branch. This resolved commit’s
+  hash is then used in the changelog entry, so the script can find and exclude
+  it from the next version’s entry.
+
+  This cannot be done via the github API and thus relies on a local clone of the
+  react-native repo.
+
+You can read more of the old documentation about it
```
[the remark this answers is not in the document]
Comment (kelset) [reply packages/rn-changelog-generator/README.md:43]: The important parts are already copy/pasted into the README itself :) The old file won't disappear either since the repo will just get archived once we merge this PR basically.

File: README.md [hunk 11]
```diff
@@ -135,6 +135,7 @@ individually, as features are added and fixes are made.
 | [@rnx-kit/metro-service](https://github.com/microsoft/rnx-kit/tree/main/packages/metro-service)                                                         | Metro service for bundling and bundle-serving                                                    |
 | [@rnx-kit/metro-swc-worker](https://github.com/microsoft/rnx-kit/tree/main/packages/metro-swc-worker)                                                   | Metro transform worker that uses swc under the hood                                              |
 | [@rnx-kit/react-native-test-app-msal](https://github.com/microsoft/rnx-kit/tree/main/packages/react-native-test-app-msal)                               | Microsoft Authentication Library (MSAL) module for react-native-test-app                         |
+| [@rnx-kit/rn-changelog-generator](https://github.com/microsoft/rnx-kit/tree/main/packages/rn-changelog-generator)                                       | New package called rn-changelog-generator                                                        |
```
Comment (afoxman) [inline README.md:138]: You should run yarn update-readme once more to get this description update (if you took my change-request for package.json description).

Comment (kelset) [reply README.md]: good shout! thanks!
